### PR TITLE
refactor: make `ApplicationContext.user` non-optional

### PR DESCRIPTION
**Pull request by Dorukyum**
## Summary
At least one of the `member` and `user` fields in the interaction payload will be present unless it's a PING interaction. This change prevents type issues from occuring almost everywhere `Interaction.user` or `ApplicationContext.user` is used without a type check.

This will be a breaking change in case this attribute is accessed in a PING handler (such a case isn't in the lib) even though it would always be `None`.

## Information
- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [x] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...).

## Checklist
- [x] I have searched the open pull requests for duplicates.
- [ ] If code changes were made then they have been tested.
  - [ ] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why.

### CODE REVIEW

**Review by Lulalaby — APPROVED** [review]
lgtm

File: discord/interactions.py [hunk 1]
```diff
@@ -103,8 +103,8 @@ class Interaction:
         The channel ID the interaction was sent from.
     application_id: :class:`int`
         The application ID that the interaction was for.
-    user: Optional[Union[:class:`User`, :class:`Member`]]
-        The user or member that sent the interaction.
+    user: Union[:class:`User`, :class:`Member`]
+        The user or member that sent the interaction. Will not be present in PING interactions.
```
Comment (BobDotCom) [inline discord/interactions.py:107]:
```suggestion
    user: Optional[Union[:class:`User`, :class:`Member`]]
        The user or member that sent the interaction. Will be `None` in PING interactions.
```

File: discord/interactions.py [hunk 2]
```diff
@@ -173,28 +173,29 @@ def _from_data(self, data: InteractionPayload):
 
         self._message_data = message_data
 
-        self.user: Optional[Union[User, Member]] = None
         self._permissions: int = 0
```
Comment (BobDotCom) [inline discord/interactions.py:176]:
```suggestion
        self.user: Optional[Union[User, Member]] = None
        self._permissions: int = 0
```

File: discord/interactions.py [hunk 3]
```diff
@@ -173,28 +173,29 @@ def _from_data(self, data: InteractionPayload):
...
-                    self.user = guild._get_and_update_member(member, int(member["user"]["id"]), cache_flag)
+        if self.type is not InteractionType.ping:
+            self.user: Union[User, Member]
+            # TODO: there's a potential data loss here
+            if self.guild_id:
+                guild = self.guild or self._state._get_guild(self.guild_id) or Object(id=self.guild_id)
+                try:
+                    member = data["member"]
+                except KeyError:
+                    pass
```
Comment (BobDotCom) [inline discord/interactions.py:186]:
```suggestion
        # TODO: there's a potential data loss here
        if self.guild_id:
            guild = self.guild or self._state._get_guild(self.guild_id) or Object(id=self.guild_id)
            try:
                member = data["member"]  # type: ignore
            except KeyError:
                pass
            else:
                self._permissions = int(member.get("permissions", 0))
                if not isinstance(guild, Object):
                    cache_flag = self._state.member_cache_flags.interaction
                    self.user = guild._get_and_update_member(member, int(member["user"]["id"]), cache_flag)
```

File: discord/interactions.py [hunk 4]
```diff
@@ -173,28 +173,29 @@ def _from_data(self, data: InteractionPayload):
...
+                    if not isinstance(guild, Object):
+                        cache_flag = self._state.member_cache_flags.interaction
+                        self.user = guild._get_and_update_member(member, int(member["user"]["id"]), cache_flag)
+                    else:
+                        self.user = Member(state=self._state, data=member, guild=guild)
+            else:
+                try:
+                    self.user = User(state=self._state, data=data["user"])
+                except KeyError:
+                    pass
```
Comment (BobDotCom) [inline discord/interactions.py:198]:
```suggestion
                    self.user = Member(state=self._state, data=member, guild=guild)
        else:
            try:
                self.user = User(state=self._state, data=data["user"])
            except KeyError:
                pass
```

File: discord/commands/context.py [hunk 5]
```diff
@@ -196,13 +196,13 @@ def message(self) -> Optional[Message]:
         return self.interaction.message
 
     @cached_property
-    def user(self) -> Optional[Union[Member, User]]:
+    def user(self) -> Union[Member, User]:
         """Union[:class:`.Member`, :class:`.User`]: Returns the user that sent this context's command.
         Shorthand for :attr:`.Interaction.user`.
         """
         return self.interaction.user
```
Comment (BobDotCom) [inline discord/commands/context.py:203]:
```suggestion
        return self.interaction.user  # type: ignore
```

**Review by BobDotCom — CHANGES_REQUESTED** [review]
Alternatively I feel it would be better to keep `Interaction.user` optional (since it is), and instead make `ApplicationContext.user` non-optional with a `type: ignore`.

**Comment by Dorukyum** [comment]
> Alternatively I feel it would be better to keep `Interaction.user` optional (since it is), and instead make `ApplicationContext.user` non-optional with a `type: ignore`.

Why keep `Interaction.user` the same? Not to remove the tons of type errors that occur in components or `on_interaction`? If we're changing one attribute, I think we should change the other too here.

**Comment by BobDotCom** [comment]
> > Alternatively I feel it would be better to keep `Interaction.user` optional (since it is), and instead make `ApplicationContext.user` non-optional with a `type: ignore`.
> 
> Why keep `Interaction.user` the same? Not to remove the tons of type errors that occur in components or `on_interaction`? If we're changing one attribute, I think we should change the other too here.

Because we should not make invalid type hints, which is what this would currently do. We can set it as optional and `type: ignore` when we know it will be present. Otherwise it defeats the purpose of having type hints, if they're wrong anyways.

**Comment by Dorukyum** [comment]
> > > Alternatively I feel it would be better to keep `Interaction.user` optional (since it is), and instead make `ApplicationContext.user` non-optional with a `type: ignore`.
> 
> > 
> 
> > Why keep `Interaction.user` the same? Not to remove the tons of type errors that occur in components or `on_interaction`? If we're changing one attribute, I think we should change the other too here.
> 
> 
> 
> Because we should not make invalid type hints, which is what this would currently do. We can set it as optional and `type: ignore` when we know it will be present. Otherwise it defeats the purpose of having type hints, if they're wrong anyways.

It'll only be wrong internally when PONG-ing interactions, if done

**Comment by Lulalaby** [comment]
ping pong

**Comment by EmreTech** [comment]
> Alternatively I feel it would be better to keep `Interaction.user` optional (since it is), and instead make `ApplicationContext.user` non-optional with a `type: ignore`.

I do not agree with this approach. It's a way better idea to just change the type hints. It'll be better in the long run.
It would also be a fantastic idea to split interactions into different classes like Disnake did, but that's for another PR.

**Comment by BobDotCom** [comment]
> I do not agree with this approach. It's a way better idea to just change the type hints. It'll be better in the long run.

Why? In the solution I proposed, `Interaction.user` is type-hinted to `Optional[User]` which granted in most cases is a bit loose as it's typically not `None`. However, we cannot say that it is never `None`, because sometimes it is. That being said, we *can* say that it is never `None` for app command interactions, hence why `ApplicationContext` would not have an optional typehint.

> It would also be a fantastic idea to split interactions into different classes like Disnake did, but that's for another PR.

That sounds like a potentially better approach. If we go this route, the best course of action may be closing this PR in favor of a new PR for the split-interaction approach.

**Comment by Dorukyum** [comment]
Splitting interaction classes is a major change and will most likely be breaking. It might just as well create more type issues like the ones this pr aims to solve.

**Comment by Dorukyum** [comment]
I've made changes after we decided to only make `ApplicationContext.user` non-optional and keep `Interaction.user` as-is.

**Review by Lulalaby — APPROVED** [review]

**Review by BobDotCom — APPROVED** [review]

**Pull request merged**